lib/src/package_analysis/lower_bound_constraint_analysis.dart

```diff
@@ -0,0 +1,378 @@
+import 'package:analyzer/dart/analysis/results.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/exception/exception.dart';
+import 'package:collection/collection.dart';
+import 'package:pana/src/package_analysis/shapes.dart';
+import 'package:path/path.dart' as path;
+import 'package:pub_semver/pub_semver.dart';
+import 'package:source_span/source_span.dart';
+
+import 'common.dart';
+
+/// Given a target *dummy* package and the summaries of the target's dependencies
+/// at their lower bound, analyze the target package and return a List of any
+/// found issues - where a symbol usage cannot be found in the relevant
+/// dependency's PackageShape
+Future<List<LowerBoundConstraintIssue>> lowerBoundConstraintAnalysis({
+  required PackageAnalysisContext context,
+  required Map<String, PackageShape> dependencySummaries,
+}) async {
+  var astVisitor = _LowerBoundConstraintVisitor(
+    context: context,
+    dependencySummaries: dependencySummaries,
+  );
+
+  final libPath = path.join(context.targetPackagePath!, 'lib');
+  final libFolder = context.folder(libPath);
+
+  // retrieve the paths of all the dart library files in this package via the
+  // resourceProvider (.dart files in ./lib)
+  final dartLibFiles = getAllFiles(libFolder)
+      .where((file) => path.extension(file.path) == '.dart')
+      .map((file) => file.path)
+      .sorted();
+
+  for (final filePath in dartLibFiles) {
+    final result = await context.analysisSession.getResolvedUnit(filePath);
+    if (result is ResolvedUnitResult) {
+      astVisitor.visitCompilationUnit(result.unit);
+    } else {
+      context.warning(
+          'Attempting to get a resolved unit for file $filePath resulted in an invalid result.');
+    }
+  }
+
+  return astVisitor.issues.values.whereNotNull().toList();
+}
+
+class _LowerBoundConstraintVisitor extends GeneralizingAstVisitor {
+  /// The [PackageAnalysisContext] corresponding to the package being analyzed.
+  final PackageAnalysisContext context;
+
+  /// The [Element] definition being analyzed.
+  late Element element;
+
+  /// The id of [element].
+  late int elementId;
+
+  // The name of the package which defines [element].
+  late String packageName;
+
+  // The summary of the package which defines [element].
+  late PackageShape dependencyShape;
+
+  /// The parent of [element].
+  late Element parentElement;
+
+  /// The [SourceSpan] corresponding to the invocation currently being analyzed.
+  late SourceSpan span;
```

Comment:
Don't it make more sense to make this a `FileSpan` for the current file (and rename it `sourceFile`).

```dart
sourceFile = SourceFile.fromString(
  context.readFile(context.uriToPath(currentUri)!),
  url: currentUri,
)
```

Then later you just do:
```dart
issues[elementId]?.references.add(sourceFile.span(identifier.offset, identifier.end));
```

Reply:
Ah, this is populated by `processIdentifier`, in that case you might want to document it.. this is slightly confusing. Another option is to return values from `processIdentifier`.